drv.go

```diff
@@ -137,6 +137,174 @@ type connPool struct {
+        return nil, errors.Errorf("initCommonCreateParams: %w", d.getError())
+    }
+    commonCreateParams.createMode = C.DPI_MODE_CREATE_DEFAULT | C.DPI_MODE_CREATE_THREADED
+    if P.EnableEvents {
+        commonCreateParams.createMode |= C.DPI_MODE_CREATE_EVENTS
+    }
+    commonCreateParams.encoding = cUTF8
+    commonCreateParams.nencoding = cUTF8
+    commonCreateParams.driverName = cDriverName
+    commonCreateParams.driverNameLength = C.uint32_t(len(DriverName))
+
+    var poolCreateParams C.dpiPoolCreateParams
+    if C.dpiContext_initPoolCreateParams(d.dpiContext, &poolCreateParams) == C.DPI_FAILURE {
+        return nil, errors.Errorf("initPoolCreateParams: %w", d.getError())
+    }
+    poolCreateParams.minSessions = DefaultPoolMinSessions
+    if P.MinSessions >= 0 {
+            poolCreateParams.minSessions = C.uint32_t(P.MinSessions)
+    }
+    poolCreateParams.maxSessions = DefaultPoolMaxSessions
+    if P.MaxSessions > 0 {
+        poolCreateParams.maxSessions = C.uint32_t(P.MaxSessions)
+    }
+    poolCreateParams.sessionIncrement = DefaultPoolIncrement
+    if P.SessionIncrement > 0 {
+        poolCreateParams.sessionIncrement = C.uint32_t(P.SessionIncrement)
+    }
+    poolCreateParams.getMode = C.DPI_MODE_POOL_GET_TIMEDWAIT
+    poolCreateParams.timeout = C.uint32_t(DefaultSessionTimeout / time.Second)
+    if P.SessionTimeout > 0 {
+        poolCreateParams.timeout = C.uint32_t(P.SessionTimeout / time.Second) // seconds before idle pool sessions get evicted
+    }
+    poolCreateParams.waitTimeout = C.uint32_t(DefaultWaitTimeout / time.Millisecond)
+    if P.WaitTimeout > 0 {
+        poolCreateParams.waitTimeout = C.uint32_t(P.WaitTimeout / time.Millisecond) // milliseconds to wait for a session to become available
+    }
+    poolCreateParams.maxLifetimeSession = C.uint32_t(DefaultMaxLifeTime / time.Second)
+    if P.MaxLifeTime > 0 {
+        poolCreateParams.maxLifetimeSession = C.uint32_t(P.MaxLifeTime / time.Second) // maximum time in seconds till a pooled session may exist
+    }
+
+    var dp *C.dpiPool
+    if Log != nil {
+        Log("C", "dpiPool_create", "username", P.UserName, "DSN", P.DSN, "common", commonCreateParams, "pool", fmt.Sprintf("%#v", poolCreateParams))
+    }
+    if C.dpiPool_create(
+            d.dpiContext,
+            cUserName, C.uint32_t(len(P.UserName)),
+            cPassword, C.uint32_t(len(P.Password)),
+            cDSN, C.uint32_t(len(P.DSN)),
+            &commonCreateParams,
+            &poolCreateParams,
+            (**C.dpiPool)(unsafe.Pointer(&dp)),
+    ) == C.DPI_FAILURE {
+        return nil, errors.Errorf("params=%v %w", P, d.getError())
+    }
+    C.dpiPool_setStmtCacheSize(dp, 40)
+    d.mu.Lock()
+    d.pools[poolKey] = &connPool{dpiPool: dp}
+    d.mu.Unlock()
+    return d.pools[poolKey], nil
+}
+
+
+func (c *conn) acquireFromPool(pool *connPool, P *ConnParams) error {
```

Comment:
Why not a method on *connPool ?
Seems a little bit strange to declare, half-initialize a connection, then acquire it from the pool.

Reply:
That could be done. The issue is that there is a method already called acquireConn() which accepts ConnectionParams, and this does not include anything to deal with sharding. I created a new structure, but ideally there should only be one. That's a bit more invasive, however, so direction would be appreciated. I'm willing to make the changes if I know what sort you would like to see.

Reply:
Yes, call it acquireConn2 for now - I can't see the whole picture yet, so leave the code merge for the future.

Reply:
Ok. I'll take care of that early next week.

Reply:
Method renamed as requested. I'll look into the merging of the two methods this week.